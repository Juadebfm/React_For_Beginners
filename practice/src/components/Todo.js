import React from "react";

function Todo({ todo }) {
  const { id, title, completed } = todo;
  const h1 = <h1 className="font-semibold font-serif">Todo - u{title.toUpperCase()}</h1>;
  const text = completed ? <strike>{h1}</strike> : h1;
  return (
    <div data-testid={`todo-${id}`} className="p-10 bg-slate-200">
      {text}
    </div>
  );
}

export default Todo;
