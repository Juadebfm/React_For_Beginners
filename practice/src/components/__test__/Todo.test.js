// import 3 things from the react test library
import { render, screen, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
// import the component
import Todo from "../Todo";

// this ensures all the test is cleaned up after run so at rerun it starts from it's original starting point. There no extra load/attachments to the components :::: remember clean up in useEffect?
afterEach(() => {
  cleanup();
});

// Below are manual testing. We can use a snapshot to makesure the component hasn't chnaged since we last run the test
test("should render non-completed Todo component", () => {
  // create todo item
  const todo = { id: 1, title: "code the app component", completed: false };
  //render the component
  render(<Todo todo={todo} />);
  //   get the id
  const todoElement = screen.getByTestId("todo-1");
  //make assertions
  expect(todoElement).toBeInTheDocument();
  //advanced
  expect(todoElement).toHaveTextContent("CODE THE APP COMPONENT");
  // check if it's contains the "<strike>" tag
});
test("should render completed Todo component", () => {
  // create todo item
  const todo = { id: 2, title: "copy the app component", completed: true };
  // render the component
  render(<Todo todo={todo} />);
  // get the todo element by test id
  const todoElement = screen.getByTestId("todo-2");
  // make assertions
  expect(todoElement).toBeInTheDocument();
  // advanced
  expect(todoElement).toHaveTextContent("COPY THE APP COMPONENT");
  // check for completion using the strike tag
  const strikeElement = todoElement.querySelector("strike");
  expect(strikeElement).toBeInTheDocument();
});

// in this one we will use snapshot which is more efficient
test("matches snapshot", () => {
  // create todo item
  const todo = { id: 2, title: "copy the app component", completed: true };
  const tree = renderer.create(<Todo todo={todo} />).toJSON();
  console.log(tree);
  expect(tree).toMatchSnapshot();
});
