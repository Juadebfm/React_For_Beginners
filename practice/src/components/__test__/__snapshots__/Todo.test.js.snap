// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="p-10 bg-slate-200"
  data-testid="todo-2"
>
  <strike>
    <h1
      className="font-semibold font-serif"
    >
      Todo - u
      COPY THE APP COMPONENT
    </h1>
  </strike>
</div>
`;
